3.2 backports 202501

vadz opened the pull request.

vadz committed:
Allow setting the exit code when returning false from OnInit()

Unfortunately returning false from the (overridden) wxApp::OnInit() has
a side effect of exiting the application with inconsistent error code:
255 under Unix but 127 under MSW with MSVC.

While changing the default exit code in this case is probably a bad idea
because there are surely people relying on the current values, at least
allow changing it by explicitly calling the new SetErrorExitCode()
function.

Unlike the version in master, this commit uses a global variable for the
exit code value to avoid changing the ABI.

(cherry picked from commit e2de7dc625d66cc1b3d336868a82c45ab8fa516f)

vadz committed:
Allow wxDataViewCustomRenderer::GetSize() return 0 in wxGTK

Returning 0 width even once from the overridden GetSize() function in
wxDataViewCustomRenderer prevented the height returned by it from being
taken into account at all by the GtkTreeView, probably due to some weird
caching going on inside GTK.

Work around this by ensuring that both size components are always
strictly positive.

See #24227, #24831.

(cherry picked from commit d3d2620273cf535f096436d16a805114bda84fe2)

dsa-t committed:
Fix wxInfoBar layout when there is not enough space for button

Ensure that we always show the "Close" button, truncating the text if
there is not enough space for both of them.

(cherry picked from commit ec99e88dbc65e1acb0c056055e10ad63ec1961b0)

vadz committed:
Ellipsize info bar text instead of truncating it

If there is not enough space to show the text in full, prefer to
ellipsize it in the middle instead of just truncating it.

Also show some text too long to fit into the window in the dialogs
sample to demonstrate that ellipsization really works.

(cherry picked from commit 3f5bbe541b99725bc6be1c9fe6d88771a1cfe4df)

vadz committed:
Simplify setting the label text in wxInfoBarGeneric

No real changes, just use the existing SetLabelText() instead of
reproducing it.

(cherry picked from commit 57bde429f441ad7fcd2512cdd72bbb1535551091)

vadz committed:
Revert "Wrap label if necessary in wxInfoBar in wxGTK"

This reverts commit 32d8b5954a1e6ff07581f0e5b55d37cad0947b6c which was
broken in both the generic and GTK version: generic one didn't work at
all and GTK one didn't adjust the height of the infobar correctly.

See #1443, #14121.

(cherry picked from commit 42882760e5e3920575588eb0bb5583a395490bbd)

vadz committed:
Also show the text of the infobar message in a tooltip

This can be useful if the entire message doesn't fit into the available
space in the infobar itself.

(cherry picked from commit 06da2d255574d51623554222b829fc137c9bec63)

vadz committed:
Run macOS CI builds using macos-14 GitHub image

"macos-12" one is deprecated and will be removed soon.

(cherry picked from commit 8d234aa5e9de1e63d80dc9d230b7b53678e64f41)

sz5000 committed:
Allow setting wxInfoBar colours before creating it

Don't override the colours if they had been already set.

See #24902.

(cherry picked from commit 247dbc57c64a1b62a85dd466a8553473d93daf24)

vslavik committed:
Fully disassociate HWND in OnEndSession

It is important to also remove the HWND from HWND-to-wxWindow mapping,
because ~wxWindow won't do it after setting its HWND to null. Otherwise
wxWndProc may crash trying to access a window that was already deleted.

Closes #24903.

(cherry picked from commit 09b6dad2ce647dec0867b020f2d687bfef862774)

vadz committed:
Keep object potentially used by libcurl alive until after cleanup

It is apparently possible to curl_multi_cleanup() to generate callbacks
that could try to use wxWebSessionCURL::m_socketPoller, so we can't
destroy this object before performing the cleanup.

Just postpone deleting it until immediately afterwards.

See #24969.

(cherry picked from commit fef7f793da0b24a8627d0d7e566f55d86dcbe8f6)

vadz committed:
Update copyright years to 2025 in 3.2 branch too

Just run misc/scripts/inc_year and commit the results.

See #18690.

lanurmi commented:
I wonder if e2de7dc625d66cc1b3d336868a82c45ab8fa516f could be backported without breaking ABI. Would be useful.

vadz commented:
It can, if we make the exit code global — which should be fine, I think.

vadz committed:
Fix multiple errors in the version script

Embarrassingly, almost none of the entries in the version script were
taken into account due to the function specifications not matching
anything in the generated library because they either omitted the
function signature entirely (which is apparently not allowed in extern
"C++" blocks) or specified it with the wrong "const" placement: "West
const" convention must be used here to match anything.

Fix the signatures to make things really work. Note that some of them
still don't have any effect because they apply to inline functions, but
it seems harmless to keep them.

PBfordev committed:
Improve window sizing documentation

Fix the flag name in wxSizerFlags::Shaped() description.

Add a reference to Windows Sizing Overview to Window Layout page and
Sizers Overview.

Add a reference to Sizers Overview to Window Sizing Overview.

Closes #25044.

(cherry picked from commit e0b055e53bbc5da9328844bb0fb2c3ac614d813f)

paulcor committed:
Fix border size for wxStaticBox with GTK3

The GtkFrame border was not being accounted for in GetBordersForSizer().

See #25045.

(cherry picked from commit 12bb1123eb620c6541593465e3a628b2a4b2faa8)

vadz committed:
Use PCRE2 function for counting the number of captures

Don't try parsing the regex ourselves, which is not just slower but also
more buggy: in particular, we didn't handle regexes with trailing
backslash correctly and could keep reading after the end of the buffer
in this case.

See #25040.

(cherry picked from commit c55ba785d3a86b9f15e30a59ab421128c6cfcf0d)

vadz committed:
Use native GTK colours for wxInfoBarGeneric with GTK 3.24

This makes the generic implementation, which can be useful even in
wxGTK, as it allows adding custom controls to the info bar, more similar
to the native one in appearance and also avoids unfortunate clashes
between the tooltip colours and the native colours of other controls,
such as hyperlinks.

See #25049, #25048.

(cherry picked from commit 015ccc7018865adbadc603f8013669d6917816fc)

MaartenBent committed:
Backport miscellaneous CMake improvements from master

- Add more configuration options for CMake installation.
- Only install the headers of the used platform/toolkit.
- Fix some bugs and warnings.

See #24792.

(cherry picked from commit 4eae795f7dd88cd94925e0c3bf56bb796904fca8)

vadz committed:
Document recent wxInfoBar backports in the change log

lanurmi committed:
Fix warnings when using event table macros with Clang 19.1.6

Avoid warnings about incompatible casts:

In file included from ../../tests/allheaders.cpp:369:
In file included from ../../tests/testprec.h:5:
In file included from ../../include/wx/evtloop.h:13:
../../include/wx/event.h:141:12: error: cast from 'void (wxEvtHandler::*)(wxFocusEvent &)' to 'wxEventFunction' (aka 'void (wxEvtHandler::*)(wxEvent &)') converts to incompatible function type [-Werror,-Wcast-function-type-mismatch]
  141 |     return reinterpret_cast<wxEventFunction>(func);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This warning was already suppressed on GCC; just do the same thing for
Clang also.

See #25064.

(cherry picked from commit 6561ca020048de57ec28fec5b27f80b00d445cdd)

PBfordev committed:
Recognize Windows Server 2025 in wxGetOsDescription()

Recognize Windows Server 2025 in wxGetOsDescription()
and update Windows OS table in wxGetOsVersion() docs.

See #25066.

(cherry picked from commit c88c86ece9b5b29a8223c9799c20aa986b455b8d)

KyrietS committed:
Prevent SDL from capturing SIGINT

Let it behave in the usual way, e.g. terminating the application by
default, instead of being effectively ignored due to being handled by
SDL and turned into an SDL event which is never processed by wxWidgets.

See #25067.

(cherry picked from commit c48c6db96d68f3c11e9db905453ba33f3fd93b22)

daute committed:
Make wxKill(wxKILL_CHILDREN) kill children recursively in wxMSW

We need to call wxKill() on children with the same wxKILL_CHILDREN flag
to kill grandchildren etc too.

See #25069, #25070.

(cherry picked from commit f2762a79f0528e4985d8c843cb1bfe54e643239b)

vadz committed:
Update copyright year in the dialogs sample too

Add it to the list of files updated by inc_year script and also change
the copyright to start at 1992 and not 1998 to make things simpler.

See #25041.

(cherry picked from commit 5252bd3ca9210edac6ae016b7fc66aa2b29896a4)

kryksyh committed:
Fix wxWindowDisabler not starting modal session under macOS

If BeginModalSession() is called twice on a new window object located at
the same address (i.e. allocated on stack), the second call takes a
shortcut due to modal window being still set to the same value and does
not start a modal session. Reset m_modalWindow once session is ended.

See #25087.

(cherry picked from commit e4285b4a35db3887347b60ce35e16a172df6c0a3)

vadz committed:
Don't remove the old file before renaming new one to it

Rename replaces the existing file, so it is not necessary to remove it
before calling it and this is actually dangerous and may result in data
loss if renaming fails, as the original file contents is just lost then.

See #25088.

(cherry picked from commit 6f861d376be2938285d6ec6fca710b5514e8459d)

vadz committed:
Use Win32 ReplaceFile() to implement wxRenameFile(overwrite=true)

This function should be atomic, unlike our own code doing copy+remove if
rename fails (as it always will under MSW if the destination exists).

(cherry picked from commit 02e6ad25605b2f6e322c13a1895d2533afcb8ac3)

Wlodzimierz-ABX-Skiba committed:
Make wxIMPLEMENT_WX_THEME_SUPPORT empty for wxUniv console apps

This makes wxIMPLEMENT_APP() work for console applications too and
fixes the build of "archive" sample when using wxUniv.

Closes #25086.

Closes #25090.

(cherry picked from commit 73c2d44f3fadedf85c732abae3c949323d7a0338)

Wlodzimierz-ABX-Skiba committed:
Don't define wxHAS_NATIVE_WINDOW in wxUniv/MSW.

wxNativeWindow is not implemented in wxUniv (even though it probably
should be), so defining wxHAS_NATIVE_WINDOW resulted in link errors
after just including this header.

This fixes widgets sample build in wxUniv/MSW.

Closes #25092.

(cherry picked from commit 05618e7fe2408a3aad7b739e5dc01680dce1f477)

vadz committed:
Add mailmap entries for Wlodzimierz Skiba

Map emails used for old (2007) and latest commits to the same identity.

(cherry picked from commit 4a9383b9ae3b2a780894c393d1d4d4c850103a32)

vadz commented:
I've accidentally pushed fixups for this branch into 3.2 instead of here, sorry for the confusion...
